change-validation/src/main/scala/com/normation/plugins/changevalidation/NotificationService.scala

```diff
@@ -0,0 +1,234 @@
+import com.normation.eventlog.EventLog
+import com.normation.rudder.domain.eventlog.AddChangeRequest
+import com.normation.rudder.domain.eventlog.DeleteChangeRequest
+import com.normation.rudder.domain.eventlog.ModifyChangeRequest
+import com.normation.rudder.domain.workflows.ChangeRequest
+import com.normation.rudder.domain.workflows.ChangeRequestId
+import com.normation.rudder.services.eventlog.ChangeRequestEventLogService
+import com.normation.rudder.web.components.DateFormaterService
+import net.liftweb.common.EmptyBox
+import net.liftweb.common.Full
+
+case class Email(value: String)
+
+case class Username(value: String)
+
+case class SMTPConf(
+    smtpHostServer : String
+  , port           : Int
+  , email          : Email
+  , login          : Option[Username]
+  , password       : Option[String]
+)
+
+case class EmailConf(
+  state     : WorkflowNode
+  , to      : Set[Email]
+  , replyTo : Set[Email]
+  , cc      : Set[Email]
+  , bcc     : Set[Email]
+  , subject : String
+  , template: String
+)
+
+
+class NotificationService(
+  path: Option[String]
+  , changeRequestEventLogService: ChangeRequestEventLogService
+) {
+  private[this] val path            = RudderConfig.RUDDER_EMAIL_CONF
+  val logger          = NamedZioLogger("plugin.change-validation")
+
+  def sendNotification(text: String, step: WorkflowNode, cr: ChangeRequest): ZIO[Any, Nothing, Object] = {
+    path match {
+      case Some(p) =>
+        val notif = for {
+          serverConfig <- getSMTPConf(p)
+          emailConf <- getStepMailConf(step, p)
+          emailBody <- getContentFromTemplate(emailConf)
+          sendingMail <- sendEmail(serverConfig, emailBody, emailConf)
+        } yield (sendingMail)
+
+        notif.catchAll {
+          err =>
+            for {
+              _ <- logger.error(s"Sending `${step.id.value}` notification have failed, cause by : ${err}")
+            } yield {
+              step
+            }
+        }
+      case None =>
+        logger.warn(s"No notification will be sent, cause by : missing email path in configuration.")
+        step
+    }
+  }
+
+  private[this] def sendEmail(conf: SMTPConf, content: String, mailParameter: EmailConf) = {
+
+    val prop = new Properties()
+    prop.put("mail.smtp.host", conf.smtpHostServer)
+    prop.put("mail.smtp.port", conf.port)
+    prop.put("mail.smtp.starttls.enable", "true")
+
+    val session = (conf.login, conf.password) match {
+      case (Some(l), Some(p)) =>
+        prop.put("mail.smtp.auth", "true");
+        val auth = new Authenticator() {
+          override protected def getPasswordAuthentication = new PasswordAuthentication(l.value, p)
+        }
+        Session.getInstance(prop, auth)
+      case (_, None)          =>
+        prop.put("mail.smtp.auth", "false");
+        Session.getInstance(prop, null)
+      case (None, _)          =>
+        prop.put("mail.smtp.auth", "false");
+        Session.getInstance(prop, null)
+    }
+    try {
```

Comment:
This should be in a IOResult.effect instead of a try

```
IOResult.effect ( "An error occured while sending email notification" ) {
  val message =
  ...
  Transport.send(message)
}
```